keras_nlp/models/bert.py

```diff
@@ -0,0 +1,287 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Bert model and layer implementations."""
+
+import tensorflow as tf
+from tensorflow import keras
+
+from keras_nlp.layers import PositionEmbedding
+from keras_nlp.layers import TransformerEncoder
+
+CLS_INDEX = 0
+
+
+class Bert(keras.Model):
+    """Bi-directional Transformer-based encoder network.
+
+    This network implements a bi-directional Transformer-based encoder as
+    described in "BERT: Pre-training of Deep Bidirectional Transformers for
+    Language Understanding" (https://arxiv.org/abs/1810.04805). It includes the
+    embedding lookups and transformer layers, but not the masked language model
+    or classification task networks.
+
+    Args:
+        vocab_size: The size of the token vocabulary.
+        num_layers: The number of transformer layers.
+        hidden_size: The size of the transformer hidden layers.
+        num_heads: The number of attention heads for each transformer.
+            The hidden size must be divisible by the number of attention heads.
+        intermediate_dim: The output dimension of the first Dense layer in a
+            two-layer feedforward network for each transformer.
+        dropout: Dropout probability for the Transformer encoder.
+        max_sequence_length: The maximum sequence length that this encoder can
+            consume. If None, max_sequence_length uses the value from sequence
+            length. This determines the variable shape for positional
+            embeddings.
+        num_segments: The number of types that the 'segment_ids' input can
+            take.
+
+    Example usage:
+    ```python
+    # Randomly initialized Bert encoder
+    encoder = keras_nlp.models.Bert(
+        vocab_size=30522,
+        num_layers=12,
+        hidden_size=768,
+        num_heads=12,
+        intermediate_dim=3072,
+        dropout=0.1,
+        max_sequence_length=12
+    )
+
+     # Call encoder on the inputs.
+    input_data = {
+        "input_ids": tf.random.uniform(
+            shape=(1, 12), dtype=tf.int64, maxval=30522),
+        "segment_ids": tf.constant(
+            [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)),
+        "input_mask": tf.constant(
+            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0], shape=(1, 12)),
+    }
+    output = encoder(input_data)
+    ```
+    """
+
+    # TODO(bischof): add tests
+
+    def __init__(
+        self,
+        vocab_size,
+        num_layers,
+        hidden_size,
```

Comment:
Should we call this `hidden_dim` and leave `hidden_dim` and `intermediate_dim` right next to each other in the list?

Reply:
`dim` is what we use in most places.

Could the distinction between "hidden" and "intermediate" be made clearer in the names?

Reply:
Left as TODO for now b/c we need to reconcile with the `TransformerEncoder` and `TransformerDecoder`

Reply:
We could go with `model_dim` and `intermediate_dim`. That would fit with "d model" which is how attention is all you need referred to this hyperparameter in a transformer architecture.

`model_dim` and `intermediate_dim` feel a little more distinct that `hidden_dim` and `intermediate_dim`.